File: generate_topic_labels.R
library(tidyverse)
library(stringr)

# Path to topic info files  
base_path <- "data/bertopic_exports/per_question/"

# Function to extract terms from KeyBERT representation
extract_keybert_terms <- function(keybert_str) {
  # Parse KeyBERT field which is a string with pairs like "'term1', 'term2'"
  terms <- str_extract_all(keybert_str, "'([^']+)'")[[1]]
  # Remove the quotes
  terms <- str_remove_all(terms, "'")
  return(terms[1:4])  # Get top 4 terms
}

# Function to create a meaningful label
create_label <- function(terms) {
  if(length(terms) == 0) return("Uncategorized")
  
  # Filter out very short or common terms
  terms <- terms[nchar(terms) > 2]
  
  if(length(terms) == 0) return("General response")
  
  # Take top 2-3 terms and create label
  label <- paste(terms[1:min(3, length(terms))], collapse = " + ")
  return(label)
}

# Collect all topics
all_topics_list <- list()

for(q_num in 1:11) {
  file_path <- glue::glue("{base_path}q{q_num}/topic_info.csv")
  
  if(file.exists(file_path)) {
    # Read just the needed columns
    data <- read_csv(file_path, col_types = cols(
      Topic = col_character(),
      Count = col_character(),
      KeyBERT = col_character(),
      .default = col_skip()
    ))
    
    # Create labels
    labels_df <- data %>%
      mutate(
        Question = paste0("Q", q_num),
        Topic_ID = as.numeric(Topic),
        Response_Count = as.numeric(Count),
        Label = sapply(KeyBERT, create_label)
      ) %>%
      select(Question, Topic_ID, Response_Count, Label) %>%
      arrange(Topic_ID)
    
    all_topics_list[[q_num]] <- labels_df
  }
}

# Combine into single dataframe
all_topics <- bind_rows(all_topics_list)

# Print formatted output
cat("\n")
cat(strrep("=", 100), "\n")
cat("BERTOPIC LABELS BY QUESTION\n")
cat(strrep("=", 100), "\n\n")

for(q in paste0("Q", 1:11)) {
  q_data <- all_topics %>% filter(Question == q) %>% arrange(Topic_ID)
  
  if(nrow(q_data) > 0) {
    cat(q, "\n")
    cat(strrep("-", 100), "\n")
    
    for(i in 1:nrow(q_data)) {
      row <- q_data[i, ]
      
      if(row$Topic_ID == -1) {
        marker <- "⊗ OUTLIERS    "
      } else {
        marker <- sprintf("  Topic %2.0f    ", row$Topic_ID)
      }
      
      cat(sprintf("%s (n=%3.0f):  %s\n", marker, row$Response_Count, row$Label))
    }
    cat("\n")
  }
}

# Save to CSV
output_df <- all_topics %>%
  mutate(
    Topic_ID = as.integer(Topic_ID),
    Response_Count = as.integer(Response_Count)
  )

write_csv(output_df, "data/topic_labels.csv")

cat(strrep("=", 100), "\n")
cat("Labels saved to: data/topic_labels.csv\n")
cat(strrep("=", 100), "\n\n")

File: generate_topic_labels.py
import pandas as pd
import re
import json
from pathlib import Path

base_path = Path("data/bertopic_exports/per_question/")

def extract_terms(representation_str):
    """Extract terms from BERTopic representation field"""
    if pd.isna(representation_str):
        return []
    # Remove brackets and split by comma, then clean up quotes
    terms = re.findall(r"'([^']+)'", str(representation_str))
    return terms[:5]  # Get top 5 terms

def create_label(terms):
    """Create a human-readable label from topic terms"""
    if not terms:
        return "Uncategorized"
    
    # Filter out overly common/generic terms
    filtered = [t for t in terms[:4] if len(t) > 2 and t not in 
                ['like', 'think', 'just', 'people', 'results', 'data', 'make', 'way']]
    
    if not filtered:
        filtered = terms[:3]
    
    return " + ".join(filtered)

# Collect all topic labels
all_topics = []

for q_dir in sorted(base_path.glob("q*")):
    topic_file = q_dir / "topic_info.csv"
    
    if topic_file.exists():
        q_num = q_dir.name.replace("q", "")
        df = pd.read_csv(topic_file)
        
        for _, row in df.iterrows():
            topic_id = int(row['Topic'])
            count = int(row['Count']) if pd.notna(row['Count']) else 0
            
            # Try each representation method in order of preference
            representation = row.get('KeyBERT', row.get('Representation', ''))
            
            terms = extract_terms(representation)
            label = create_label(terms)
            
            all_topics.append({
                'Question': f"Q{q_num}",
                'Topic': topic_id,
                'Count': count,
                'Top_Terms': " + ".join(terms[:4]) if terms else "N/A",
                'Label': label
            })

# Create output dataframe and save
output_df = pd.DataFrame(all_topics).sort_values(['Question', 'Topic'])
output_df.to_csv("data/topic_labels.csv", index=False)

# Print formatted output
print("\n" + "="*100)
print("BERTopic LABELS BY QUESTION")
print("="*100)

for q in [f"Q{i}" for i in range(1, 12)]:
    q_data = output_df[output_df['Question'] == q]
    if len(q_data) > 0:
        print(f"\n{q}")
        print("-"*100)
        for _, row in q_data.iterrows():
            topic_marker = "⊗ OUTLIERS" if row['Topic'] == -1 else f"  Topic {int(row['Topic']):2d}"
            print(f"{topic_marker:15s} (n={int(row['Count']):3d}):  {row['Label']}")

print("\n" + "="*100)
print(f"Complete labels saved to: data/topic_labels.csv")
print("="*100 + "\n")

# Also create a more detailed version
detailed_df = output_df[['Question', 'Topic', 'Count', 'Top_Terms', 'Label']]
detailed_df.columns = ['Question', 'Topic_ID', 'Response_Count', 'Top_Terms', 'Topic_Label']
detailed_df.to_csv("data/topic_labels_detailed.csv", index=False)
print(f"Detailed version saved to: data/topic_labels_detailed.csv\n")
